### PR TITLE
feat: adds new health check field

### LINKED ISSUE #4183: Allow users to specify http paths in HTTPS and HTTP control-plane health checks
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

Users should be able to specify an HTTP or HTTPS path for their control-plane load balancers from AWS. Currently, we only allow the user to specify the health check [protocol for a classic ELB](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/5538c89bdfe245afa76bd9b72c0f64c92a865325/api/v1beta2/network_types.go#L59). We would like a more comprehensive check by hitting the `/healthz` endpoint which kube-apiserver exposes to the load balancer. 

Furthermore, users should be allowed to modify this [healthcheck field](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/5538c89bdfe245afa76bd9b72c0f64c92a865325/api/v1beta2/awscluster_webhook.go#L123) because this is a field that AWS allows you to configure health checks on classic loadbalancers. 

A comprehensive list of changes I'd like to see:

1. Support users passing in paths 
2. Allow users to update healhcheckprotocol 
3. Change the name of healthcheckprotocol to healthcheck to more accurately describe allowing paths 

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api-provider-aws version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by faiq**
**What type of PR is this?**

**What this PR does / why we need it**:
/kind feature

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4183 

**Special notes for your reviewer**:

**Checklist**:

- [x] squashed commits
- [] includes documentation
- [x] adds unit tests
- [] adds or updates e2e tests

**Release note**:

```release-note
feat: adds new health check field
```

### CODE REVIEW

File: api/v1beta2/elb.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+func (e *ELBHealthCheckTarget) ValidateELBHealthCheck() []*field.Error {
+	if e == nil {
+		return nil
+	}
+	var errs field.ErrorList
+	re := regexp.MustCompile(`^(TCP|SSL|HTTP|HTTPS|TLS|UDP):(\d+)\/?(\w*)$`)
```
Comment (Skarlso) [inline api/v1beta2/elb.go:30]: Build this outside so it's not rebuilt on each Validate call.

Comment (faiq) [reply api/v1beta2/elb.go:30]: Good idea!

File: api/v1beta2/awscluster_types.go [hunk 2]
```diff
@@ -198,11 +198,17 @@ type AWSLoadBalancerSpec struct {
 	Subnets []string `json:"subnets,omitempty"`
 
 	// HealthCheckProtocol sets the protocol type for ELB health check target
+	// This field will be deprecated in a future API use HealthCheck
```
Comment (Skarlso) [inline api/v1beta2/awscluster_types.go:201]:
```suggestion
	// Deprecated: This field will be deprecated in a future API use HealthCheck
```

File: api/v1beta2/awscluster_types.go [hunk 3]
```diff
@@ -198,11 +198,17 @@ type AWSLoadBalancerSpec struct {
...
 	// default value is ELBProtocolSSL
 	// +kubebuilder:validation:Enum=TCP;SSL;HTTP;HTTPS;TLS;UDP
 	// +optional
 	HealthCheckProtocol *ELBProtocol `json:"healthCheckProtocol,omitempty"`
 
+	// HealthCheckTarget sets a ELB health check target
```
Comment (Skarlso) [inline api/v1beta2/awscluster_types.go:207]:
```suggestion
	// HealthCheckTarget sets an ELB health check target
```

**Comment by faiq** [comment]
Another idea: what if we `/readyz` whenever the user sets the protocol to `HTTPS` or `HTTP` this way we can avoid adding a new field in the API Spec.

**Comment by Skarlso** [comment]
@faiq Err, is that an append to this or separate from it? :D

**Comment by faiq** [comment]
@Skarlso the PR is a separate way of solving the problem

**Comment by Skarlso** [comment]
Sorry. Does that mean that this pr is obsolete then? Or is it just an alternative approach and we should decide which one we take?

**Comment by faiq** [comment]
Yep! its an alternative approach and the maintainers could choose what direction we should go in. Both work and are ready. We just need to pick an approach

**Comment by Skarlso** [comment]
Well, I would go for the one with less code in it. :D

**Comment by faiq** [comment]
closing in favor of #4227

**Pull request closed without merging**